### ISSUE TITLE
Default/shared filesystem paths for config & storage

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#3956]
## The Problem

The Zed/Brim/Brimcap tools as they've existed to date have caused user confusion due to how they rely on a combination of default paths and/or environment variables to locate configuration, state, and storage data on which they rely. Some examples:

* After a user has installed/launched the Brim app, an app-managed Zed lake is stored in a platform-specific directory (paths described [here](https://github.com/brimdata/brim/wiki/Filesystem-Paths#user-data)). If they later install/run standalone Zed CLI commands, they would need to create the `ZED_LAKE` environment variable pointed at this same path or localhost network endpoint if they want their app and CLI experiences to converge on the same lake.

* The same problem occurs in reverse if the user has first run the Zed CLI tools to create/query a Zed lake, then they install the Brim app: They're likely to end up with two wholly separate lakes.

* A similar set of problems also exist for config state, such as the cached record of the most recent pool/branch selection made with `zed use`.

* In order to manage its Zed lake, the Brim app has always shipped with its own Zed binaries that live in yet another platform-specific directory (paths described [here](https://github.com/brimdata/brim/wiki/Filesystem-Paths#application-binaries)). If the user is unaware of the presence/location of these embedded binaries, they may install redundant Zed CLI commands, which wastes storage and has the potential to create versioning issues if the "a la carte" CLI commands are out-of-sync with the lake storage format as of the app's ship date.

* Once again in reverse, if the user already had the Zed CLI tools installed and in their `PATH`, they are exposed to a similar problem if they later install the Brim app: Redundant binaries and possible versioning problems.

* A similar set of problems exist for the "Brimcap root", as the Brim app traditionally ships with an embedded `brimcap` binary and manages a packet index in own "Brimcap root", but the Brimcap tools could also be installed/run "a la carte" outside the app and potentially point to the same/different Brimcap root (https://github.com/brimdata/brimcap/issues/154). For simplification, the remainder of this issue will speak only of the Zed lake example, with the assumption that a similar approach could be applied to solving the problem for Brimcap if we choose.

## Goals

In brainstorming solutions within the Dev team, there's been general consensus around a few goals.

1. No tool/app should require users to set environment variables before storing/querying data. By default, different tools that use a Zed lake should find/leverage the same lake in well-known locations.

2. Zed CLI commands should no longer default to attempting to find a local lake on a localhost service endpoint, but rather should rely on default filesystem paths.

3. Users should not need to reference detailed documents to find storage/config files on their filesystem, and these paths should be platform-independent whenever possible. One example would be short pathnames relative to a user's "home directory".

4. Customized overrides to these paths (such as via environment variables for CLI tooling or prompts in a GUI installer for an app) should be possible, but most users are expected to stick with defaults, so the majority of effort should on sensible default behavior.

5. If choosing a convention to follow, behavior can lean more toward behaving like developer tools (e.g., `git`) rather than desktop apps (e.g., a photo editor).

6. To allow tools to safely use shared storage/config and binaries, strict implementation and adherence to Zed versioning conventions should be fully implemented as part of addressing this problem. This will allow notification of users when upgrades are necessary.

## Precedent

In seeking inspiration to solve this problem in a way that may click with users, the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) is worthy of study. In brief, it attempts to solve this problem for Linux desktop apps similarly to how it's believed to have ben addressed for macOS ([Apple's docs](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/FileSystemOverview/FileSystemOverview.html#//apple_ref/doc/uid/TP40010672-CH2-SW1), and [XDG comparison](https://stackoverflow.com/questions/3373948/equivalents-of-xdg-config-home-and-xdg-data-home-on-mac-os-x)) and [Windows](https://stackoverflow.com/questions/43853548/xdg-basedir-directories-for-windows).

It's worthy of note that, probably due to its use of Electron, the Brim app has ended up following at least some of the XDG convention on Linux, since running the app results in the creation of a `$HOME/.config/Brim` path, which could also be described as `$XDG_CONFIG_HOME/Brim`.

```
$ ls -l ~/.config/Brim
total 80
-rw-rw-r-- 1 phil phil  3773 May 17 14:41  appState.json
drwx------ 3 phil phil  4096 May 17 14:31  blob_storage
drwx------ 3 phil phil  4096 May 17 14:38  Cache
drwx------ 4 phil phil  4096 May 17 14:31 'Code Cache'
-rw------- 1 phil phil 20480 May 17 14:31  Cookies
-rw------- 1 phil phil     0 May 17 14:31  Cookies-journal
drwx------ 2 phil phil  4096 May 17 14:31 'Crash Reports'
drwxrwxr-x 4 phil phil  4096 May 17 14:31  data
drwx------ 2 phil phil  4096 May 17 14:31  Dictionaries
drwx------ 2 phil phil  4096 May 17 14:31  GPUCache
drwxr-xr-x 4 phil phil  4096 May 17 14:31  lake
drwx------ 3 phil phil  4096 May 17 14:31 'Local Storage'
drwx------ 2 phil phil  4096 May 17 14:31  logs
-rw------- 1 phil phil   750 May 17 14:32 'Network Persistent State'
-rw------- 1 phil phil    57 May 17 14:31  Preferences
drwx------ 2 phil phil  4096 May 17 14:38 'Session Storage'
-rw------- 1 phil phil   203 May 17 14:31  TransportSecurity
```

That said, this does not wholly match with what XDG advocates, as the specification advocates the use of a separate `$XDG_DATA_HOME` for data files, hence Brim's use of the `lake/` and `data/` directories somewhere below `~/.config` goes against convention.

Once again, XDG only seeks to solve this problem for Linux. While the links provided above show developers following their instincts to try to apply the same approach on macOS/Windows and (hence leverage a universal cross-platform convention), consensus response seems to point users back at the traditional Apple/Microsoft-provided conventions, which leads back to the long/ugly paths we wanted to avoid in our 2nd goal above. That said, if we apply our 4th goal from above about leaning toward developer-centric behavior, some precedent appears such as Git's global `.gifconfig` that's assumed to live below the user's home directory even on macOS and [Windows](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/Where-system-global-and-local-Windows-Git-config-files-are-saved) (though note the [Git docs](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration) also mention `~/.config/git/config` as an alternative, as if making room for the XDG-style behavior) or [NPM configstore](https://www.npmjs.com/package/configstore), which similarly relies on "`$XDG_CONFIG_HOME` or `~/.config`", hence a `.config` path below the user's home directory on both macOS and Windows.

Note a "home directory" is only assumed to be defined in a `$HOME` env var on Linux and macOS, whereas the Windows equivalent is [`%USERPROFILE%`](https://en.wikipedia.org/wiki/%25USERPROFILE%25`https://en.wikipedia.org/wiki/Home_directory`).

**Comment by mattnibs** [comment]
@philrz I'm not an XDG expert but I've seen other projects following XDG place config in  `$HOME/.config` and data in `$HOME/.local`.

**Comment by philrz** [comment]
@mattnibs: Yes, what you describe is what I saw in the spec, so it makes sense that you've seen other projects do the same.  I was trying to highlight two things in the examples I cited:

1.  The fact that Brim has always ended up putting config under `~/.config` happens to line up with XDG convention, but the fact Brim puts _data storage_ below there does not.
2.  The way I saw Git or NPM configstore use `~/.config` on macOS and Windows happens to line up with XDG convention, even though XDG is only described as targeted at Linux.

**Comment by philrz** [comment]
A community user recently mentioned in a separate issue https://github.com/brimdata/brim/issues/2371 the desire to customize the _installation_ path, which is separate from the config/storage paths described above, but perhaps this should be rolled into the same effort. In their original wording:

> Windows platform：
> Hope to customize the installation path or provide a portable installation package, easy to put on the U disk can be used anywhere (currently the installed files moved to other paths can be used, no problems found at the moment)

**Comment by philrz** [comment]
Another community user asked in a [Slack thread](https://brimdata.slack.com/archives/CTSMAK6G7/p1670333882916119) about customizing the location for lake storage.

> I have two drives on my PC. One is C:\ (OS) and the other is M:\ (storage). Is there a config file or other means to point Zui to a lake on the M:\ drive, outside of the default `C:\Users\me\AppData\Roaming\Zui - Insiders`? The only way I know to do this at the moment is with `zed server -lake <filepath>`. Trying to see if there's a way to do this without having to start the zed service.

I was able to confirm an effective one-off workaround by creating a [directory junction](https://www.howtogeek.com/howto/16226/complete-guide-to-symbolic-links-symlinks-on-windows-or-linux/) to. However, a more direct/documented approach would surely have surely helped.

**Comment by philrz** [comment]
Another community Zui user asked in https://github.com/brimdata/zui/issues/2753 about wanting to make some customizations to where their lake data is stored. In their own words:

> Changing the storage location of pools
> 
> In some materials where the data is very large, for example 20-30 GB, currently all data is stored in the following folder.
The problem of lack of hard disk space occurs
> `C:\Users*\AppData\Roaming\Zui\lake`
> 
> Is it possible to store lake data in another location? Possibility of optional selection for the storage location of each pool

For now I've pointed them at the same "directory junction" trick cited in the previous comment. Note that they implied a desire to maybe change storage location on a pool-by-pool basis rather than just whole lake.

**Comment by philrz** [comment]
@mccanne recently once again confronted the aspects of this issue related to lake storage. Specifically, he already had some data he'd loaded into Zed lakes using the `zed` CLI commands, then he started Zui in Dev mode hoping that it would automatically find his existing lake data but saw that it did not. He was aware he could start `zed serve -lake ...` & specify his pre-existing storage path before starting Zui, but this is not his preference.

During group brainstorming on the topic, @mattnibs proposed & established consensus on an approach he's seen in other tools: `zed serve` could look for lakes in an ordered set of storage locations, including (but not necessarily limited to) the XDG & common OS-convention paths such as those described above in this issue. Since Zui already starts a `zed serve` process, this would have made @mccanne's use case work as expected as long as the CLI tooling/docs had steered him toward one of the storage locations that the `zed serve` started behind Zui would have found. For the case of a new user that has no prior lake storage, no lakes would be found among the ordered set of storage locations and therefore the current Zui behavior would still apply where the `zed serve` would initialize an empty lake at one of these locations. Also, an env variable (e.g., `ZED_LAKE`) could still override the check of all these paths, if set.
